### PR TITLE
Refactor/relaxed compare (depends on #138 MPI)

### LINKED ISSUE #92: consolidate relaxed comparisons
ASGarD
    asgardasgard
    Issues
    #92

Open
Opened 4 days ago by McDaniel, Benjamin T
Close issue
New issue
consolidate relaxed comparisons

we have these lambdas littered across the test code base. need to unify them in tests_general perhaps.
Related issues
0

Related merge requests
1

    Feature/improved split
    asgard/asgard !83

McDaniel, Benjamin T @3bm added hygeine label 4 days ago
McDaniel, Benjamin T @3bm mentioned in merge request !83 (merged) 4 days ago

Graham Lopez
Graham Lopez @bgl · 4 days ago
Owner

it would also make the tests way easier to read on failures if we could find a way to use Catch2's Approx()

Graham Lopez @bgl assigned to @7qm 4 days ago

### PR DESCRIPTION
**Pull request by bmcdanie**
closes #92 .

### CODE REVIEW

File: testing/tests_general.hpp [hunk 1]
```diff
@@ -74,3 +75,43 @@ std::string to_string_with_precision(T const a_value, int const precision = 6)
...
+// compare two templated container types.
+// scaled for value magnitude
+// tol_fac can be used to adjust tolerance; this number is multipled by epsilon
+// to form the tolerance
+template<typename comparable_1, typename comparable_2>
+void relaxed_comparison(comparable_1 const &first, comparable_2 const &second,
+                        double const tol_fac = 1e1)
+{
+  REQUIRE(first.size() == second.size());
+  using P = typename std::remove_pointer<decltype(first.data())>::type;
```
Comment (grahamlopez) [inline testing/tests_general.hpp]: let's tell the reader why this is needed

File: testing/tests_general.hpp [hunk 2]
```diff
@@ -74,3 +75,43 @@ std::string to_string_with_precision(T const a_value, int const precision = 6)
...
+  auto first_it                    = first.begin();
+  std::for_each(
+      second.begin(), second.end(),
+      [&first_it, tol_fac](auto const &second_elem) {
+        auto const tol = std::numeric_limits<P>::epsilon() * tol_fac;
+        auto const scale_fac =
+            std::max(std::max(static_cast<P>(1.0), std::abs(*first_it)),
+                     std::abs(second_elem));
+        REQUIRE_THAT(*first_it++,
+                     Catch::Matchers::WithinAbs(second_elem, tol * scale_fac));
```
Comment (grahamlopez) [inline testing/tests_general.hpp]: @dlg0 how do you feel about this, in the context of our previous discussions?

File: src/batch_tests.cpp [hunk 3]
```diff
@@ -6,6 +6,8 @@
 #include "tests_general.hpp"
 #include <numeric>
 #include <random>
+
+auto const tol_scale = 1e4;
```
Comment (grahamlopez) [inline src/batch_tests.cpp]: why? how was this chosen?
why are you setting it globally for the file, when it is used in one spot?

File: src/time_advance_tests.cpp [hunk 4]
```diff
@@ -20,24 +20,9 @@ struct distribution_test_init
...
-      std::abs(*std::max_element(diff.begin(), diff.end(), abs_compare));
-
-  P const tolerance = std::numeric_limits<P>::epsilon() *
-                      (std::is_same<P, double>::value ? 1e5 : 1e3);
-  REQUIRE(result <= tolerance);
-}
-
 int const num_steps          = 5;
 int const workspace_limit_MB = 1000;
+auto const tol_scale         = 1e4;
```
Comment (grahamlopez) [inline src/time_advance_tests.cpp]: same questions as for `batch_tests.cpp`
why? how was this chosen?
why are you setting it globally for the file, when it is used in one spot?

File: src/basis_tests.cpp [hunk 5]
```diff
@@ -5,17 +5,10 @@
 #include <numeric>
 #include <random>
 
+auto const tol_scale = 1e3;
```
Comment (grahamlopez) [inline src/basis_tests.cpp]: okay, at least this global is used in more than one spot.
maybe name it something like `basis_tests_tolerance` to indicate its scope?

File: src/coefficients_tests.cpp [hunk 6]
```diff
@@ -35,7 +18,7 @@ TEMPLATE_TEST_CASE("continuity 1 (single term)", "[coefficients]", double,
   fk::matrix<TestType> const test =
       continuity1->get_coefficients(0, 0).clone_onto_host();
 
-  relaxed_comparison<TestType>(gold, test, 1e2);
+  relaxed_comparison(gold, test, 1e2);
```
Comment (grahamlopez) [inline src/coefficients_tests.cpp]: (picking one of these at random)
we need a way to explain and justify these magic values (in all of the testing files). Minimally, should we add a comment to the top of the file saying they were chosen empirically and last vetted as of some date? I don't know a good solution here.
I'm opening this for discussion. (paging @dlg0 as well)

File: src/pde_tests.cpp [hunk 7]
```diff
@@ -4,11 +4,7 @@
 #include "tests_general.hpp"
 #include <vector>
 
-// our trig functions don't exactly match matlab. we need a little wiggle room.
-auto relaxed_compare = [](auto fx, auto gold) {
-  auto relaxed_epsilon = std::numeric_limits<decltype(gold)>::epsilon() * 1e2;
-  REQUIRE(Approx(fx).epsilon(relaxed_epsilon) == gold);
-};
+auto const tol_scale = 1e2;
```
Comment (grahamlopez) [inline src/pde_tests.cpp]: rinse and repeat comments from elsewhere. how chosen? rename... etc.

File: src/quadrature_tests.cpp [hunk 8]
```diff
@@ -4,23 +4,10 @@
 #include "matlab_utilities.hpp"
 #include "tests_general.hpp"
 
+auto const tol_scale = 1e2;
```
Comment (grahamlopez) [inline src/quadrature_tests.cpp]: rinse and repeat comments from elsewhere. how chosen? rename... etc.

File: src/transformations_tests.cpp [hunk 9]
```diff
@@ -5,6 +5,7 @@
 #include "transformations.hpp"
 #include <numeric>
 
+auto const tol_scale = 1e3;
```
Comment (grahamlopez) [inline src/transformations_tests.cpp]: rinse and repeat comments from elsewhere. how chosen? rename... etc.

File: testing/tests_general.hpp (hunk 1 above)
Comment (bmcdanie) [reply testing/tests_general.hpp]: comment added

File: src/batch_tests.cpp (hunk 3 above)
Comment (bmcdanie) [reply src/batch_tests.cpp]: moved into lowest scope, comment added

File: src/time_advance_tests.cpp (hunk 4 above)
Comment (bmcdanie) [reply src/time_advance_tests.cpp]: moved into lowest scope, comment added

File: src/basis_tests.cpp (hunk 5 above)
Comment (bmcdanie) [reply src/basis_tests.cpp]: done

File: src/coefficients_tests.cpp (hunk 6 above)
Comment (bmcdanie) [reply src/coefficients_tests.cpp]: added comments about how chosen

File: src/pde_tests.cpp (hunk 7 above)
Comment (bmcdanie) [reply src/pde_tests.cpp]: done

File: src/quadrature_tests.cpp (hunk 8 above)
Comment (bmcdanie) [reply src/quadrature_tests.cpp]: done

File: src/transformations_tests.cpp (hunk 9 above)
Comment (bmcdanie) [reply src/transformations_tests.cpp]: done

File: testing/tests_general.hpp (hunk 2 above)
Comment (grahamlopez) [reply testing/tests_general.hpp]: moving on. I'm sure that we'll continue to discuss

**Pull request merged**